### PR TITLE
Update to net50 and fix warnings

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/JamesNK/Newtonsoft.Json/issues/2423

### CODE REVIEW

File: Src/Newtonsoft.Json/Linq/JContainer.cs [hunk 1]
```diff
@@ -273,7 +273,7 @@ public override JEnumerable<JToken> Children()
         /// <returns>
         /// A <see cref="IEnumerable{T}"/> containing the child values of this <see cref="JToken"/>, in document order.
         /// </returns>
-        public override IEnumerable<T> Values<T>()
+        public override IEnumerable<T?> Values<T>() where T : default
```
Comment (TylerBrinkley) [inline Src/Newtonsoft.Json/Linq/JContainer.cs:276]: I'm not familiar with this constraint. Was this added in C# 9?

File: Src/Newtonsoft.Json/Linq/JToken.cs [hunk 2]
```diff
@@ -334,7 +334,7 @@ public IEnumerable<JToken> BeforeSelf()
         /// <typeparam name="T">The type to convert the token to.</typeparam>
         /// <param name="key">The token key.</param>
         /// <returns>The converted token value.</returns>
-        public virtual T Value<T>(object key)
+        public virtual T? Value<T>(object key)
```
Comment (TylerBrinkley) [inline Src/Newtonsoft.Json/Linq/JToken.cs:337]: Is adding nullability annotations to generic type arguments now supported in C# 9? I didn't see anything related to that in the blog posts or in the what's new in C# 9 guide at https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9.

Comment (0xF6) [reply Src/Newtonsoft.Json/Linq/JToken.cs:337]: This need backward compatibility, need use directives #nullable and #if NET50?

File: Src/Newtonsoft.Json/Linq/JContainer.cs (hunk 1 above)
Comment (JamesNK) [reply Src/Newtonsoft.Json/Linq/JContainer.cs:276]: Yes https://github.com/dotnet/csharplang/blob/9807f9b2e2105255c0a658f3afc908769bb31339/proposals/csharp-9.0/unconstrained-type-parameter-annotations.md#default-constraint

File: Src/Newtonsoft.Json/Linq/JToken.cs (hunk 2 above)
Comment (JamesNK) [reply Src/Newtonsoft.Json/Linq/JToken.cs:337]: > Is adding nullability annotations to generic type arguments now supported in C# 9? I didn't see anything related to that in the blog posts or in the what's new in C# 9 guide at [docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9).

https://github.com/dotnet/csharplang/blob/9807f9b2e2105255c0a658f3afc908769bb31339/proposals/csharp-9.0/unconstrained-type-parameter-annotations.md

File: Src/Newtonsoft.Json/Linq/JContainer.cs (hunk 1 above)
Comment (TylerBrinkley) [reply Src/Newtonsoft.Json/Linq/JContainer.cs:276]: @JamesNK Thanks for the reference, I appreciate it!

**Pull request merged**